app/services/EnsureGuaranteeService.scala

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+import models.ParseError.{AdditionalInfoCodeMissing, AdditionalInfoMissing, GuaranteeAmountZero, GuaranteeNotFound, GuaranteeTypeInvalid, InvalidItemNumber, MissingItemNumber, NoGuaranteeReferenceNumber, SpecialMentionNotFound}
+import models.{ChangeGuaranteeInstruction, GooBlock, Guarantee, NoChangeGuaranteeInstruction, NoChangeInstruction, ParseError, ParseHandling, SpecialMention, SpecialMentionGuarantee, SpecialMentionGuaranteeDetails, SpecialMentionOther, TransformInstruction, TransformInstructionSet}
+import cats.data.ReaderT
+import cats.implicits._
+import com.google.inject.Inject
+import utils.guaranteeParsing.{GuaranteeXmlReaders, InstructionBuilder}
+
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+import scala.xml.{Elem, Node, NodeSeq}
+import scala.xml.transform.{RewriteRule, RuleTransformer}
+
+class EnsureGuaranteeService @Inject()(xmlReaders: GuaranteeXmlReaders, instructionBuilder: InstructionBuilder) extends ParseHandling {
+
+
+  def ensureGuarantee(xml: NodeSeq): ParseHandler[NodeSeq] =
+    parseInstructionSets(xml) match {
+      case Left(error) => Left(error)
+      case Right(instructionSets) =>
+        Right(updateXml(prunedXml(xml), instructionSets))
+    }
+
+  def parseInstructionSets(xml: NodeSeq): ParseHandler[Seq[TransformInstructionSet]] = {
+    println("parseGuarantees"); xmlReaders.parseGuarantees(xml) match {
+      case Left(error) => Left(error)
+      case Right(guarantees) =>
+        println("gooBlock"); xmlReaders.gooBlock(xml) match {
+          case Left(error) => Left(error)
+          case Right(gooBlocks) =>
+            ParseError.liftParseError(gooBlocks.map {
+              block =>
+                println("getInstructionSet"); getInstructionSet(block, guarantees).map {
+                  instructionSet => instructionSet
+                }
+            })
+        }
+    }
+  }
+
+  def updateXml(prunedXml: NodeSeq, instructionSets: Seq[TransformInstructionSet]): NodeSeq = {
+    new RuleTransformer(new RewriteRule {
+      override def transform(node: Node): NodeSeq = {
+        node match {
+          case e: Elem if e.label == "GOOITEGDS" => {
+            xmlReaders.gooBlock(e) match {
+              case Left(_) => throw new Exception()
+              case Right(blocks) => {
+                val currentBlock = blocks.head
```

Comment:
I think this dangerous `head` might be because the parsers are wrong

Reply:
Oh, I need to ensure this is a headOpt - I'll have to confirm the examples I have are legit, but I'll revisit this code.